cmd/gaia/init/testnet.go

```diff
@@ -182,92 +200,148 @@ func testnetWithConfig(config *cfg.Config, cdc *codec.Codec) error {
 			return err
 		}
 
-		// Gather gentxs folder
+		// gather gentxs folder
 		err = writeFile(fmt.Sprintf("%v.json", nodeDirName), gentxsDir, txBytes)
 		if err != nil {
 			_ = os.RemoveAll(outDir)
 			return err
 		}
 	}
 
-	// Generate empty genesis.json
+	if err := initGenFiles(cdc, chainID, accs, genFiles, numValidators); err != nil {
+		return err
+	}
+
+	err := collectGenFiles(
+		cdc, config, chainID, monikers, nodeIDs, valPubKeys, numValidators,
+		outDir, viper.GetString(flagNodeDirPrefix), viper.GetString(flagNodeDaemonHome),
+	)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Successfully initialized %d node directories\n", numValidators)
+	return nil
+}
+
+func initGenFiles(
+	cdc *codec.Codec, chainID string, accs []app.GenesisAccount,
+	genFiles []string, numValidators int,
+) error {
+
 	appGenState := app.NewDefaultGenesisState()
 	appGenState.Accounts = accs
+
 	appGenStateJSON, err := codec.MarshalJSONIndent(cdc, appGenState)
 	if err != nil {
 		return err
 	}
+
 	genDoc := types.GenesisDoc{
 		ChainID:    chainID,
 		AppState:   appGenStateJSON,
 		Validators: nil,
 	}
 
-	// Save all genesis.json files
+	// generate empty genesis files for each validator and save
 	for i := 0; i < numValidators; i++ {
 		if err := genDoc.SaveAs(genFiles[i]); err != nil {
 			return err
 		}
 	}
 
+	return nil
+}
+
+func collectGenFiles(
+	cdc *codec.Codec, config *cfg.Config, chainID string,
+	monikers, nodeIDs []string, valPubKeys []crypto.PubKey,
+	numValidators int, outDir, nodeDirPrefix, nodeDaemonHomeName string,
+) error {
+
+	var appState json.RawMessage
+	genTime := tmtime.Now()
+
 	for i := 0; i < numValidators; i++ {
-		nodeDirName := fmt.Sprintf("%s%d", viper.GetString(nodeDirPrefix), i)
-		nodeDaemonHomeName := viper.GetString(nodeDaemonHome)
+		nodeDirName := fmt.Sprintf("%s%d", nodeDirPrefix, i)
 		nodeDir := filepath.Join(outDir, nodeDirName, nodeDaemonHomeName)
 		gentxsDir := filepath.Join(outDir, "gentxs")
 		moniker := monikers[i]
 		config.Moniker = nodeDirName
+
 		config.SetRoot(nodeDir)
 
 		nodeID, valPubKey := nodeIDs[i], valPubKeys[i]
-		// Run `init` and generate genesis.json and config.toml
 		initCfg := initConfig{
 			ChainID:   chainID,
 			GenTxsDir: gentxsDir,
 			Name:      moniker,
 			NodeID:    nodeID,
 			ValPubKey: valPubKey,
 		}
+
 		genDoc, err := loadGenesisDoc(cdc, config.GenesisFile())
 		if err != nil {
 			return err
 		}
-		if _, err := genAppStateFromConfig(cdc, config, initCfg, genDoc); err != nil {
+
+		nodeAppState, err := genAppStateFromConfig(cdc, config, initCfg, genDoc)
```

Comment:
genAppStateFromConfig was always a misnomer... it also writes the genesis file.  Then we write it again below.
Lets fix this by making genAppStateFromConfig stateless and *not* write the genesis file.